### PR TITLE
Preserve non-grayscale colors when inverting images

### PR DESCRIPTION
**Pull request by aebrer**
## Summary

Add a draw buffer `invertColors` flag so selected non-grayscale text and background colors can be pre-inverted when KOReader's night-mode image inversion path is active.

This lets colored EPUB text/backgrounds keep their intended hue after the final framebuffer inversion, matching the existing goal of preserving image colors in night mode.

## Implementation

- Adds `setInvertColors()` / `getInvertColors()` state to `LVDrawBuf`.
- Keeps this separate from `invertImages`, while allowing the frontend to drive both together.
- Applies `invertNonGrayscaleColor()` at draw time for:
  - text color drawing;
  - selected background drawing paths.
- Leaves reserved/current, fully transparent, and grayscale colors unchanged.
- Preserves alpha by XORing only RGB bits.

## Testing

Tested with a debug Android KOReader build using the frontend wiring suggested in discussion:

- `invertColors = invertImages`
- night mode enabled
- color rendering enabled
- synthetic EPUB covering colored inline text, backgrounds, table/cell backgrounds, grayscale, transparent, inherited/currentcolor, and semi-transparent cases
- real EPUB excerpts with colored inline text and table/system-box content

With image inversion off, colors still invert normally in night mode. With image inversion on, non-grayscale text/background colors are preserved as intended.

## Notes

This PR intentionally does not change SVG/image transparency behavior. That was explored separately and dropped to preserve existing image inversion behavior.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/crengine/677)

### CODE REVIEW

**Comment by aebrer** [comment]
Okay! I'm back! Way sooner than this weekend. I couldn't stop thinking about this. 


This time, I am confident that I have a working solution that is also not aesthetically gross as fuck.

Here is a .zip containing a minimal README plus two test files: 
- a .css style tweak that is meant to be used in everyday night-mode reading (ie. the one I am now using)
- a test epub that has a bunch of useful tables and colors and so on in it (reads like an optical illusion lol)

white and black text specifically are *expected* to be exempt from the double-inversion, so they will appear inverted in night-mode (intended)

I also tried to avoid the AI agent's tendencies to overcomplicated based on hypothetical edge-cases that are never likely to actually occur. I *think* it should be pretty clean and minimal at this point. 

Tested manually and quite pleased with the outcome (thanks for pointing me to cr-hints)

test files: 
[cr-hint-invert-colors-upstream-artifacts.zip](https://github.com/user-attachments/files/29024145/cr-hint-invert-colors-upstream-artifacts.zip)


 edit: btw, thanks again for the feedback on how I was contributing to open source projects like this. I tried to keep this one smaller, manually tested, and much less  noisy. I hope it lands better, and I’d like to use this approach as a template going forward.

**Comment by poire-z** [comment]
Just a quick test for now:
Without the cr-hint:
image

With it:
image

File: crengine/include/cssdef.h [hunk 1]
```diff
@@ -452,6 +452,9 @@ enum css_generic_value_t {
 // the right things to avoid any overflow.
 #define CSS_CR_HINT_NO_CAP_IMAGE_SIZE       0x00000020 // -cr-hint: no-cap-image-size
 
+// Pre-invert non-grayscale colors so night mode inverts them back to their original hue.
```
Comment (poire-z) [inline crengine/include/cssdef.h]: May be mention "...so night mode _when drawbuf.getInvertImages()  is set_ inverts them back..." ?

File: crengine/include/lvtextfm.h [hunk 2]
```diff
@@ -544,6 +544,20 @@ class LFormattedText
     ~LFormattedText() { lvtextFreeFormatter( m_pbuffer ); }
 };
 
+// Pre-invert a non-grayscale color for -cr-hint: invert-colors.
+// Preserves alpha, and leaves reserved, transparent, and grayscale values unchanged.
+inline lUInt32 crHintInvertColor(lUInt32 color)
+{
```
Comment (poire-z) [inline crengine/include/lvtextfm.h]: Feels this one is not really lvtextfm related, and would find a better place elsewhere ? May be in lvdrawbuf.h (which has more stuff about colors) or lvrend.h (which has littler helpers like getStyledImageSize, lengthToPx...).

File: crengine/src/lvtextfm.cpp [hunk 3]
```diff
@@ -42,6 +42,25 @@
 #define MAX_ADDED_LETTER_SPACING_PERCENT 0
 #define CJK_WIDTH_SCALE_PERCENT 100
 
+static lUInt32 crHintAdjustColor(LVDrawBuf * buf, const css_style_ref_t style, lUInt32 color)
```
Comment (poire-z) [inline crengine/src/lvtextfm.cpp]: `static inilne` like the other in lvrend.cpp ?

They feel a bit odd standalone and in-your-face at the top of lvtextfm.cpp while there is not a single over similar helper.

Maybe they could be moved just above its sole user LFormattedText::Draw() ? if there's no other generic location for such helpers.

**Review by poire-z — COMMENTED** [review]
> This time, I am confident that I have a working solution that is also not aesthetically gross as fuck.

Indeed, it's really not :)

**Comment by poire-z** [comment]
The integration ends up indeed quite non-intrusive and clean - and done at Draw time, on the stuff drawn only (so, probably less expensive than at style set time).

Did you have any issue doing it with setNodeStyle() - how/why did you or it decided to do it at draw time ?
I'm fine with having it non-inherited (using `* { -cr-hint: invert-color }` will get a similar effect), but why did you decide for that ?

Any reason for users (and may be us adding it to our style tweak menu) to not use  `* { -cr-hint: invert-color }` - and use a more targeted style tweak like your cr-hint-real-broad.css ?

(It obviously doesn't work well on the black text over yellow background I added to your test file. But not caring if you're happy with the general effect.)

> and I’d like to use this approach as a template going forward.

You mean "in other projects" ? Or do you have any other ideas for crengine/koreader ? If you do, please run them by us before getting your AI to work on it and opening a PR :) (We may have suggestions, or a chance to shout an early "no go":).)

**Comment by aebrer** [comment]
Thanks, this is all fair.

For the inline comments: agreed, `lvtextfm.h` was a convenient shared include but not really the right semantic home. I can move/adjust that helper so it is less in-your-face, and clarify the `getInvertImages()` coupling in the comment. Just need to do it a bit later when I have some time. 

On `setNodeStyle()`: I didn't hit a hard blocker there, but draw-time felt like the cleaner layer once I traced it. The behavior depends on the active draw buffer/night-mode image-inversion path, and I didn't want to bake a mode-dependent color into computed styles or require restyling when night mode changes. Doing it at draw time also means it only touches colors that are actually drawn, after `currentcolor`/background selection has resolved. I think. Basically it just seemed "more correct" so I went for it.

I kept the hint non-inherited mostly as the safer/minimal choice. This is more of a rendering tweak than a normal CSS property, and inheritance would make a container rule affect descendant text/background colors that the tweak did not necessarily intend to preserve. If someone wants the broad behavior, `* { -cr-hint: invert-colors }` gives that explicitly. Just figured it was more minimal that way.

About using `*`: I think it is a valid blunt option, but I would not make it the only/recommended example. It preserves all non-grayscale publisher colors, which can be nice, but it can also preserve colors in combinations that are not readable after night mode. Your black-text-on-yellow-background example is exactly that kind of limitation: black/white/grayscale are intentionally left alone, so the black text still becomes white while the yellow background is preserved. A targeted tweak is less surprising for my use case: preserve known colored boxes/tables/spans, while leaving ordinary text/night-mode behavior alone. tl;dr: I don't think it matters much. And yeah my thinking with some of the edge cases (good thinking with the black on yellow bg), was that if  it's annoying in one specific case in the book, I can toggle the style tweak just to read that rare and weird line, then toggle it back on after. In like 99% of real book cases the double inversion "just works" so I'm pretty happy with it. Main target was litrpg genre books sourced from RoyalRoad. :)

And yes, by “template going forward” I meant my general contribution workflow, not that I have a queue of crengine/KOReader PRs ready to drop :) For anything else here, I'll run the idea by you first before doing implementation work or opening a PR. I just think I should improve my general instructions for agents in case other users use it... I don't want to accidentally train up a new generation of slop posters lol, bare minimum they can have the agent encouraging them to be polite XD

**Comment by poire-z** [comment]
Was about to ask you if we really need that -cr-hint, and if we can't just do it by default when invertImages (or have out bottom toggle become a 3-state toggle, `off | invert images | invert images & colors`) as it feels it would just be normal to keep colors.

But you're saying you still prefer a -cr-hint , which allows interested users to work on styletweaks to  to target some specific elements only - and keep some bad colors on others ?
Fine by me either way, I just want to be sure we're doing the best thing.

**Comment by aebrer** [comment]
Ah no! Sorry I was focused on the cr-hint as that's what I thought you were suggesting as a first pass and for testing. IN terms of UX I think the best+simplest is to just tie it to the image toggle probably, as you have said. Alternatively a second toggle for colors, but I think the pattern of adding more and more toggles can be a bit of a slippery slope lol

I like the idea of the image toggle going between off | invert images | invert images & colors, I think that's a clean way to handle it (especially compared to a user style tweak we can expect most users to never learn about)

(or even just having it on by default with the images makes sense too. happy to work on whatever if you want me to implement a toggle too, though I'm going to wait for your feedback first before I start making changes)

**Comment by poire-z** [comment]
> on the cr-hint as that's what I thought you were suggesting as a first pass and for testing

I did, as a quick way to not be bothered by UI tweaks, and as a non-intrusive way to enable it if the feature is niche and not worth more.

> Alternatively a second toggle for colors, but I think the pattern of adding more and more toggles can be a bit of a slippery slope 

Indeed.

> I like the idea of the image toggle going between off | invert images | invert images & colors, I think that's a clean way to handle it

The wording for the toggles needs to be short, and even if we find small stuff (no idea) for English, it might be hard for translators to do the same, so let's not do that (but keep the c++ stuff able to support it).

> or even just having it on by default with the images makes sense too

So, let's go with that. (If users complain about that colors change, we will think about a solution to keep invertImages independant of invertColors).

I'd say to just add setInvertColors, getInvertColors, _invertColors just as similar to the existing ones for InvertImages in crengine/include/lvdrawbuf.h (so, the drawBuf carries this request).
Drop the cr_hint.
And instead of having `if ( buf->getInvertImages() && !style.isNull() && STYLE_HAS_CR_HINT(style, INVERT_COLORS) )`, just have `if ( buf->getInvertColors() )`

For the cre.cpp, no need to ask an AI, we human can still do little adjustments better - so just use that (for testing, don't make a PR in koreader-base with that, I'll do it when I'll bring crengine to frontend, in a few days with possibly other stuff):
```diff
--- a/cre.cpp
+++ b/cre.cpp
@@ -3792,6 +3792,10 @@ static int drawCurrentPage(lua_State *L) {
        if (lua_isboolean(L, 6)) {
                dithering = lua_toboolean(L, 6);
        }
+       bool invert_colors = invert_images;
+       if (lua_isboolean(L, 7)) {
+               invert_colors = lua_toboolean(L, 7);
+       }

        int w = bb->w;
        int h = bb->h;
@@ -3806,6 +3810,7 @@ static int drawCurrentPage(lua_State *L) {
                 * Blitbuffer.TYPE_BBRGB32, see CreDocument:drawCurrentView */
                LVColorDrawBuf drawBuf(w, h, bb->data, 32);
                drawBuf.setInvertImages(invert_images);
+               drawBuf.setInvertColors(invert_colors);
                drawBuf.setSmoothScalingImages(smooth_scaling);
                doc->text_view->Draw(drawBuf, false);
                drawn_images_count = drawBuf.getDrawnImagesCount();
```

**Comment by aebrer** [comment]
Updated as suggested!

Changes:
- dropped the -cr-hint flag and parser branch
- added drawbuf-level setInvertColors/getInvertColors state, mirroring invertImages
- switched the color adjustment path to use drawbuf.getInvertColors()
- kept this crengine-only; I did not open a koreader-base PR

I tested locally with the cre.cpp wiring you provided, using invert_colors = invert_images.

With Invert Images off, colors still invert normally in night mode. With Invert Images on, the non-grayscale text/background colors are preserved as intended.

The test EPUB title in the screenshots still says cr-hint from the earlier test artifact, but this run is using the updated drawbuf-level invertColors path.

signal-2026-06-18-175507
signal-2026-06-18-175507_002

**Comment by poire-z** [comment]
Looks quite simple and neat :)
May be you can even get rid of the far-away little helpers (I hate it that AI always add such helpers, in spite of my instructions to not do that) with something like:
```diff
- bg_color = adjustInvertedColor(drawbuf, bg_color);
+ bg_color = drawbuf.getInvertColors()  ? invertNonGrayscaleColor(color) : bg_color;
```

**Comment by aebrer** [comment]
Done! Good suggestion thanks. Rebuilt and tested, still works exactly as expected.

**Review by poire-z — APPROVED** [review]
Approving the current stuff (way simpler than your initial idea, I like when it goes that way :)).
We may wait a bit before merging and bumping it into frontend (we'llsee if other stuff is coming about our new FontManager).

May be you'll get something as simple to be added to this PR for transparent SVG.

**Comment by aebrer** [comment]
Added the SVG-specific piece as a separate commit, linked from the discussion in koreader/koreader#15520.

I also tested with my synthetic EPUB and the two SVG test EPUBs you shared there. I didn’t notice any issues: transparent SVG areas stay transparent, grayscale/currentColor SVG line art follows night mode, and non-grayscale SVG colors are preserved. Interestingly, even the full art SVG in there of the anime cafe looked good! 

The new commit is intentionally kept separate so it can be rebased/squashed however you prefer later.

**Comment by aebrer** [comment]
and yeah I can really see now how insanely overdone that first approach was, LMAO. I also like this minimal change much better, considering it's not only simpler but also "just works" lol

File: crengine/include/lvdrawbuf.h [hunk 4]
```diff
@@ -445,6 +452,19 @@ inline lUInt32 RevRGB( lUInt32 cl ) {
...
 }
 
+// Pre-invert a non-grayscale color so night mode inverts it back to its original hue.
+// Preserves alpha, and leaves reserved, transparent, and grayscale values unchanged.
+inline lUInt32 invertNonGrayscaleColor( lUInt32 color ) {
+    if ( (color & 0xFFFFFFFE) == 0xFFFFFFFE || (color & 0xFF000000) == 0xFF000000 )
+        return color;
+    lUInt32 r = color & 0x00FF0000;
+    lUInt32 g = (color << 8) & 0x00FF0000;
+    lUInt32 b = (color << 16) & 0x00FF0000;
```
Comment (NiLuJe) [inline crengine/include/lvdrawbuf.h]: Nit: I was mildly confused by the masking, this reads slightly better to me with a standard shift and truncate approach (especially since we don't actually need either the pure color itself, or the components for anything besides a comparison):

```suggestion
    const lUInt8 b = color & 0xFF;
    const lUInt8 g = (color >> 8) & 0xFF;
    const lUInt8 r = (color >> 16) & 0xFF;
```

NOTE: I don't actually remember if CRe's pixel format is RGB or BGR, so, err, variable names may also be wrong ;p.

Comment (NiLuJe) [reply crengine/include/lvdrawbuf.h]: > NOTE: I don't actually remember if CRe's pixel format is RGB or BGR, so, err, variable names may also be wrong ;p.

If I trust a comment above, drawbuf is BGRA.

But if this deals with actual CSS colors and not drawbuf colors, it's probably RGBA, so we're good?

**Review by NiLuJe — COMMENTED** [review]
Just an über-nit ;).

**Comment by poire-z** [comment]
Feels like LVImageScaledDrawCallback() is generic and not strictly about drawing SVG ?
Would this affect of rimage formats ? (jpegs photos having some gray pixels and some colors ones, becoming odd?)

**Comment by aebrer** [comment]
The new path is gated on `src->IsScalable()`, which currently means the LunaSVG image source, so JPEG/PNG/WebP/etc. should keep the existing `invertImages` behavior.                                                                                                   
                                                                                                     
I also applied NiLuJe’s readability suggestion for the component extraction.

File: crengine/include/lvdrawbuf.h (hunk 4 above)
Comment (aebrer) [reply crengine/include/lvdrawbuf.h]: If we trust the AI investigator:

> Numeric lUInt32 colors in CRe are effectively 0xAARRGGBB, with inverted alpha in the top byte.

but I guess the inversion actually works either way so it doesn't really matter much, as you say

in theory it's now correct though

(checked `crengine/src/lvstsheet.cpp` -> `parse_color_value()`, and also some comments in `crengine/src/lvimg.cpp`)

**Comment by poire-z** [comment]
Some test/observations/comments in https://github.com/koreader/koreader/issues/15520#issuecomment-4761846416 and followups.

**Comment by poire-z** [comment]
Can you ask your AI (or yourself :)) to update this PR title and description ?

**Comment by aebrer** [comment]
Done! We can't pretend that I wrote all those words by hand :laughing:, but this should now reflect the actual PR contents a lot more accurately.

**Comment by poire-z** [comment]
Was about to merge and wondered: did we forget border colors (`DrawBorder()`)?
Or the changes would be too heavy/ugly in this complex function setting lots of colors, and unnecessary ?

(Saving here my Squash commit message:
```
DrawBuf: allow drawing colors inverted (#677)

Will be requested when in night mode: inverted colors will
then get inverted back by night mode and displayed normally.
Applies only to non-grayscale text and background colors.
```
)

**Comment by aebrer** [comment]
Good catch, yes, borders were missed. I remember looking into it but it seemed more complex back when my original plan was bloated and kind of dumb. Turns out it's not that complex after all. 

I added a small follow-up commit: `837e3d9c Apply color inversion to borders`.

It covers both border drawing paths I found:

- `lvrend.cpp::DrawBorder()` for block/table borders;
- `lvtextfm.cpp::drawBorder()` for inline/text-flow borders.

For the larger `DrawBorder()` path, I kept the existing color/shading logic intact and only apply `invertNonGrayscaleColor()` after the side color and any `shadecolor`/`lightcolor` values are resolved. So groove/ridge/inset/outset should keep their existing relative shading. This was the part I was hesitating about, because it is a bit more complex than the other inversions... but tested with new stuff in my epub and it seems to work well.

I rebuilt the Android arm64 debug APK and tested with an updated EPUB containing solid/dashed/double/groove block borders, inline borders, `currentcolor` borders, table/header/cell borders, and gray controls. Looks good to me.

If you find this too much for the final squash, feel free to drop that last commit specifically

:)
